README.rst
=========================
djangocms-geoplaceholder
=========================

Плагин-контейнер для Django CMS, выводящий дочерние плагины только для
посетителей из определённых регионов, определяемых с помощью `django-geoip <https://github.com/futurecolors/django-geoip>`_.

Установка и настройка
----------------------

Проще всего установить его с помощью `pip <https://pypi.python.org/pypi/pip>`_ ::

$ pip install djangocms-geoplaceholder

Добавьте 'djangocms_geoplaceholder' в INSTALLED_APPS вашего проекта и запустите ::

$ manage.py migrate

Не забудьте обновить базу адресов django-geoip ::

$ manage.py geoip_update


CHANGES.rst
История изменений
=================

1.1.0 (2019-07-28)
------------------

* Поддержка Django 2.1 и Django CMS 3.6.


djangocms_geoplaceholder/forms.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django import forms
from django.utils.translation import ugettext as _

from django_geoip.models import City, Region, Country

from .models import GeoPlaceholderModel


class GeoPlaceholderForm(forms.ModelForm):
    country = forms.ModelChoiceField(label=GeoPlaceholderModel._meta.get_field('country').verbose_name.title(),
                                     queryset=Country.objects.order_by('name'), required=False)
    region = forms.ModelChoiceField(label=GeoPlaceholderModel._meta.get_field('region').verbose_name.title(),
                                    queryset=Region.objects.order_by('name'), required=False)
    city = forms.ModelChoiceField(label=GeoPlaceholderModel._meta.get_field('city').verbose_name.title(),
                                  queryset=City.objects.order_by('name'), required=False)

    class Meta:
        model = GeoPlaceholderModel
        exclude = []

    def clean(self):
        cleaned_data = super(GeoPlaceholderForm, self).clean()
        if sum(1 for v in cleaned_data.values() if v is None) != 2:
            raise forms.ValidationError(_('The one of country, region or city must be chosen!'))
        return cleaned_data


djangocms_geoplaceholder/models.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import models
from django.utils.encoding import python_2_unicode_compatible
from django.utils.translation import ugettext as _

from cms.models import CMSPlugin
from django_geoip.models import City, Region, Country


@python_2_unicode_compatible
class GeoPlaceholderModel(CMSPlugin):
    country = models.ForeignKey(Country, verbose_name=_('Country'), null=True, blank=True, on_delete=models.SET_NULL)
    region = models.ForeignKey(Region, verbose_name=_('Region'), null=True, blank=True, on_delete=models.SET_NULL)
    city = models.ForeignKey(City, verbose_name=_('City'), null=True, blank=True, on_delete=models.SET_NULL)

    def __str__(self):
        location = self.country or self.region or self.city
        return u'{}'.format(location)


djangocms_geoplaceholder/cms_plugins.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.utils.translation import ugettext as _

from cms.plugin_base import CMSPluginBase
from cms.plugin_pool import plugin_pool
from ipware.ip import get_ip
from django_geoip.models import IpRange

from .models import GeoPlaceholderModel
from .forms import GeoPlaceholderForm


@plugin_pool.register_plugin
class GeoPlaceholderPublisher(CMSPluginBase):
    model = GeoPlaceholderModel
    form = GeoPlaceholderForm
    name = _('Geo-targeting placeholder')
    render_template = "plugins/geo_placeholder.html"
    allow_children = True

    def render(self, context, instance, placeholder):
        context = super(GeoPlaceholderPublisher, self).render(context, instance, placeholder)

        geo_placeholder_render = False
        if instance.country:
            attr_name = 'country'
        if instance.region:
            attr_name = 'region'
        if instance.city:
            attr_name = 'city'

        request = context['request']
        remote_addr = get_ip(request)
        if remote_addr:
            try:
                geoip_record = IpRange.objects.by_ip(remote_addr)
            except IpRange.DoesNotExist:
                pass
            else:
                if getattr(geoip_record, attr_name) == getattr(instance, attr_name):
                    geo_placeholder_render = True

        context.update({'geo_placeholder_render': geo_placeholder_render})
        return context


djangocms_geoplaceholder/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.9.9 on 2017-04-24 08:15
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('cms', '0016_auto_20160608_1535'),
        ('django_geoip', '0001_initial'),
    ]

    operations = [
        migrations.CreateModel(
            name='GeoPlaceholderModel',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='geoplaceholder_geoplaceholdermodel', serialize=False, to='cms.CMSPlugin')),
                ('city', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='django_geoip.City', verbose_name='\u0413\u043e\u0440\u043e\u0434')),
                ('country', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='django_geoip.Country', verbose_name='\u0421\u0442\u0440\u0430\u043d\u0430')),
                ('region', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='django_geoip.Region', verbose_name='\u0420\u0435\u0433\u0438\u043e\u043d')),
            ],
            options={
                'abstract': False,
            },
            bases=('cms.cmsplugin',),
        ),
    ]


djangocms_geoplaceholder/migrations/0002_auto_20190728_2026.py
# Generated by Django 2.1 on 2019-07-28 12:26

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        ('djangocms_geoplaceholder', '0001_initial'),
    ]

    operations = [
        migrations.AlterField(
            model_name='geoplaceholdermodel',
            name='city',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='django_geoip.City', verbose_name='Город'),
        ),
        migrations.AlterField(
            model_name='geoplaceholdermodel',
            name='cmsplugin_ptr',
            field=models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='djangocms_geoplaceholder_geoplaceholdermodel', serialize=False, to='cms.CMSPlugin'),
        ),
        migrations.AlterField(
            model_name='geoplaceholdermodel',
            name='country',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='django_geoip.Country', verbose_name='Страна'),
        ),
        migrations.AlterField(
            model_name='geoplaceholdermodel',
            name='region',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.SET_NULL, to='django_geoip.Region', verbose_name='Регион'),
        ),
    ]
